File: 2021년/7월/0711 ABCDE.py
# 백준_13023_ABCDE_DFS_골드 5

import sys
input = sys.stdin.readline


def dfs(idx, cnt):
    global flag
    if cnt == 4:
        flag = 1
        return

    visited[idx] = 1
    for i in range(len(graph[idx])):
        if visited[graph[idx][i]] == 0:
            dfs(graph[idx][i], cnt+1)
            if flag:
                return

    visited[idx] = 0


n, m = map(int, input().split())
graph = [[] for _ in range(n)]
visited = [0]*n
flag = 0

for _ in range(m):
    x, y = map(int, input().split())
    graph[x].append(y)
    graph[y].append(x)

for i in range(n):
    dfs(i, 0)
    print(visited)
    if flag:
        break

print(flag)

File: 2021년/7월/0711 ABCDE.cpp
#include <iostream>
#include <vector>

using namespace std;

int a,b,flag = 0;
vector<int> visited;
vector<int>N[2000];

void dfs(int index, int count)
{
	if(count == 4)
	{
		flag = 1;
		return;
	}
	
	for(int i=0;i<a;i++)
	{
		if(N[index][i] == 1 and visited[i] == 0) 
		{
			visited[i] = 1;
			dfs(i,count+1);
			visited[i] = 0;
		}
	}
}

int main()
{
	int x,y;
	cin >> a >> b;
	
	for(int i=0;i<b;i++)
	{
		cin >> x >> y;
		N[x].push_back(y);
		N[y].push_back(x);
	}
	
	for(int i=0;i<a;i++)
	{
		visited[i] = 1;
		dfs(i,0);
		visited[i] = 0;
		
		if(flag == 1) break;
	}
	
	cout << flag;
	
}
